02_knight_and_princess/tevfik_aksoy/main.cpp

```diff
@@ -1,93 +1,186 @@
 #include <cassert>
+#include <ctime>
+#include <filesystem>
 #include <fstream>
 #include <iostream>
+#include <random>
 #include <string>
 #include <vector>
 
-using std::to_string;
+using std::string;
 using std::vector;
 
 using VertexId = int;
 using EdgeId = int;
+using VertexDepth = int;
 
 struct Edge;
 struct Vertex;
+
+namespace validation {
+bool edge_id_exists_in_vertex(const EdgeId& edge_id,
+                              const vector<EdgeId>& edge_ids) {
+  for (const auto& edge : edge_ids) {
+    if (edge_id == edge) {
+      return true;
+    }
+  }
+  return false;
+}
+
 bool vertex_exists_in_graph(const VertexId& id, const vector<Vertex>& vertices);
 bool edge_exists_in_graph(const EdgeId& id, const vector<Edge>& edges);
-bool edge_id_exists_in_vertex(const EdgeId& id, const vector<EdgeId>& edge_ids);
-bool is_vertex_id_valid(const VertexId& id);
-bool is_edge_id_valid(const EdgeId& id);
+
+bool is_vertex_id_valid(const VertexId& id) {
+  if (id < 0) {
+    return false;
+  }
+  return true;
+}
+
+bool is_edge_id_valid(const EdgeId& id) {
+  if (id < 0) {
+    return false;
+  }
+  return true;
+}
+}  // namespace validation
+
+bool probability(float condition) {
+  std::random_device rd;
+  std::mt19937 mt(rd());
+  std::uniform_real_distribution<float> probability(0.0, 100.0);
+  if (probability(mt) >= condition)
+    return true;
+  else
+    return false;
+}
 
 struct Vertex {
  public:
+  const VertexId source_id;
   const VertexId id;
+  const VertexDepth depth;
 
-  explicit Vertex(const VertexId& _id) : id(_id) {}
+  Vertex(const VertexId& source_id,
+         const VertexId& id,
+         const VertexDepth& depth)
+      : source_id(source_id), id(id), depth(depth) {}
 
-  void add_edge_id(const EdgeId& _id) {
-    assert(!edge_id_exists_in_vertex(_id, edge_ids_) &&
+  void add_edge(const EdgeId& _id) {
+    assert(!validation::edge_id_exists_in_vertex(_id, edge_ids_) &&
            "Edge already exists in vertex!");
     edge_ids_.push_back(_id);
   }
-
+  const vector<EdgeId>& get_edge_ids() const { return edge_ids_; }
   std::string to_JSON() const {
     std::string json_string;
 
-    json_string += "\t{ \"id\": " + to_string(id) + ", \"edge_ids\": [";
+    json_string += "\t{ \"id\": " + std::to_string(id) + ", \"edge_ids\": [";
     for (int i = 0; i < edge_ids_.size(); i++) {
-      json_string += to_string(edge_ids_[i]);
+      json_string += std::to_string(edge_ids_[i]);
       if (i + 1 != edge_ids_.size())
         json_string += ", ";
     }
+    json_string += "], \"depth\": " + std::to_string(depth) + "}";
     return json_string;
   }
 
-  const vector<EdgeId>& get_edge_ids() const { return edge_ids_; }
-
  private:
   vector<EdgeId> edge_ids_;
 };
 
 struct Edge {
  public:
+  enum class Color { GRAY, GREEN, BLUE, YELLOW, RED };

```

Comment:
Это не константы, не нужно называть их все большими буквами. Это скорее как тип данных, класс или структура:
```suggestion
  enum class Color { Gray, Green, Blue, Yellow, Red };
```
Обновил `Naming Convention` для `Enum`

Reply:
Понял, исправил